include/tbbexec/tbb_thread_pool.hpp

```diff
@@ -0,0 +1,413 @@
+            friend auto tag_invoke(stdexec::get_completion_signatures_t, Self&&, Env)
+                -> completion_signatures<Self, Env>
+                requires true
+            ;
+
+            template <stdexec::tag_category<stdexec::forwarding_sender_query> Tag, class... As>
+                requires stdexec::__callable<Tag, const Sender&, As...>
+            friend auto tag_invoke(Tag tag,
+                                   const bulk_sender& self,
+                                   As&&... as) noexcept(stdexec::__nothrow_callable<Tag, const Sender&, As...>)
+                -> stdexec::__call_result_if_t<stdexec::tag_category<Tag, stdexec::forwarding_sender_query>,
+                                               Tag,
+                                               const Sender&,
+                                               As...> {
+                return ((Tag &&) tag)(self.sndr_, (As &&) as...);
+            }
+        };
+
+        friend sender tag_invoke(stdexec::schedule_t, const scheduler& s) noexcept { return sender{*s.pool_}; }
+
+        template <stdexec::sender Sender, std::integral Shape, class Fun>
+        using bulk_sender_t =
+            bulk_sender<stdexec::__x<std::remove_cvref_t<Sender>>, Shape, stdexec::__x<std::remove_cvref_t<Fun>>>;
+
+        template <stdexec::sender S, std::integral Shape, class Fn>
+        friend bulk_sender_t<S, Shape, Fn> tag_invoke(stdexec::bulk_t,
+                                                      const scheduler& sch,
+                                                      S&& sndr,
+                                                      Shape shape,
+                                                      Fn fun) noexcept {
+            return bulk_sender_t<S, Shape, Fn>{*sch.pool_, (S &&) sndr, shape, (Fn &&) fun};
+        }
+
+        friend stdexec::forward_progress_guarantee tag_invoke(stdexec::get_forward_progress_guarantee_t,
+                                                              const tbb_thread_pool&) noexcept {
+            return stdexec::forward_progress_guarantee::parallel;
+        }
+
+        friend class tbb_thread_pool;
+        explicit scheduler(tbb_thread_pool& pool) noexcept : pool_(&pool) {}
+
+        tbb_thread_pool* pool_;
+    };
+
+    scheduler get_scheduler() noexcept { return scheduler{*this}; }
+
+    void request_stop() noexcept { /* TODO? */
+    }
+
+    std::uint32_t available_parallelism() const { return m_arena.max_concurrency(); }
+
+private:
+    void enqueue(task_base* task) noexcept {
+        m_arena.enqueue([task] { tbb::this_task_arena::isolate([task] { task->__execute(task, /*tid=*/0); }); });
+    }
+    void bulk_enqueue(task_base* task, std::uint32_t n_threads) noexcept {
+        for (std::size_t i = 0; i < n_threads; ++i) {
+            // FIXME: Is this right? i is the tid?
+            m_arena.enqueue(
+                [task, i] { tbb::this_task_arena::isolate([task, i] { task->__execute(task, /*tid=*/i); }); });
```

Comment:
My hope as a user who cares about performance, is that the implementation wouldn't need to worry about work stealing (the original Cilk term, which I prefer to the less descriptive "moonlighting").  Work stealing is essential to performance of not-necessarily-load-balanced parallel loops.

Does C++17 `parallel_for` need to worry about users doing funny things with `thread_local`?  If not, then I would hope that the P2300 wording would let `bulk` forbid that kind of user behavior.  Otherwise, then perhaps that's a change to consider for P2300 itself.

Reply:
@ericniebler says that we shouldn't need `isolate` here.